table/tables/partition.go

```diff
@@ -334,6 +338,126 @@ func generateRangePartitionExpr(ctx sessionctx.Context, pi *model.PartitionInfo,
 	return ret, nil
 }
 
+func generateListPartitionExpr(ctx sessionctx.Context, pi *model.PartitionInfo,
+	columns []*expression.Column, names types.NameSlice) (*PartitionExpr, error) {
+	// The caller should assure partition info is not nil.
+	locateExprs := make([]expression.Expression, 0, len(pi.Definitions))
+	schema := expression.NewSchema(columns...)
+	p := parser.New()
+	for _, def := range pi.Definitions {
+		exprStr, err := generateListPartitionExprStr(ctx, pi, schema, names, &def, p)
+		if err != nil {
+			return nil, err
+		}
+		expr, err := parseSimpleExprWithNames(p, ctx, exprStr, schema, names)
+		if err != nil {
+			// If it got an error here, ddl may hang forever, so this error log is important.
+			logutil.BgLogger().Error("wrong table partition expression", zap.String("expression", exprStr), zap.Error(err))
+			return nil, errors.Trace(err)
+		}
+		locateExprs = append(locateExprs, expr)
+	}
+	ret := &PartitionExpr{
+		InValues: locateExprs,
+	}
+	return ret, nil
+}
+
+func generateListPartitionExprStr(ctx sessionctx.Context, pi *model.PartitionInfo,
+	schema *expression.Schema, names types.NameSlice, def *model.PartitionDefinition, p *parser.Parser) (string, error) {
+	var partStr string
+	if len(pi.Columns) == 0 {
+		partStr = pi.Expr
+	} else if len(pi.Columns) == 1 {
+		partStr = pi.Columns[0].L
+	} else {
+		return generateMultiListColumnsPartitionExprStr(ctx, pi, schema, names, def, p)
+	}
```

Comment:
the structure seems strange,  how about packaging the below code as `generateSingleListColumnsPartitionExprStr` and returning them like L374 does